src/server/report/result.js

```diff
@@ -162,20 +153,27 @@ function renderResult(resultData, options) {
 }
 
 export function renderBrowserResults(browser, tests, options) {
-	const results = tests.map(t => {
+	const filteredTests = tests.filter(t => {
 		const resultData = t.results.find(r => r.name === browser.name);
-		const duration = resultData.duration;
-		const durationClass = {
-			'error': duration >= 1000,
-			'result-duration': true,
-			'pass': duration < 500,
-			'warning': duration >= 500 && duration < 1000
-		};
+		if (resultData.passed && options.filterStatus === FILTER_STATUS.FAILED ||
+			!resultData.passed && options.filterStatus === FILTER_STATUS.PASSED) {
+			return false;
+		}
+		return true;
+	});
+	const results = filteredTests.map(t => {
```

Comment:
Instead of looping through `tests` and `results` twice each, I'd use a `reduce`
```javascript
export function renderBrowserResults(browser, tests, options) {
	const results = tests.reduce((acc, t) => {
		const resultData = t.results.find(r => r.name === browser.name);
		if (resultData.passed && options.filterStatus === FILTER_STATUS.FAILED ||
			!resultData.passed && options.filterStatus === FILTER_STATUS.PASSED) {
			return acc;
		}

		let status = STATUS_TYPE.NORMAL;
		if (resultData.duration >= 1000) {
			status = STATUS_TYPE.ERROR;
		} else if (resultData.duration >= 500) {
			status = STATUS_TYPE.WARNING;
		}

		return acc.push(html`
			<div class="result-container">
				<div class="result-test-name">
					<h3>${t.name}</h3>
					<div class="result-duration">${renderStatusText(`${resultData.duration}ms`, status)}</div>
				</div>
				${renderResult(resultData, options)}
			</div>
		`) && acc;
	}, []);
	return html`
		<div class="result-browser padding">
			${ICON_BROWSERS[browser.name]}
			<div>
				<div class="result-browser-name">${browser.name}</div>
				<div>version ${browser.version}</div>
			</div>
		</div>
		${results.length === 0 ? renderEmpty() : results}
	`;
}
```

Reply:
That's a really elegant way to filter & map at the same time -- thanks!